### ISSUE TITLE
Allow multiple profile for same camera

### ISSUE THREAD

**Issue by ExcelsiorEyer** [VorlonCD/bi-aidetection#104]
Blue Iris has allowed for &PRESET to be added to the filename in its latest version 5.3.6.5. For a PTZ camera which could have multiple presets, it would be helpful to allow creation of multiple entries for same camera (as the mask would be different for each preset) when the files have the preset values included in the input files.

**Comment by tripp396** [comment]
I think multiple profiles would be helpful for this use case and a few others

**Comment by VorlonCD** [comment]
Can you provide a bit more detail how you would like the feature to work?

**Comment by ExcelsiorEyer** [comment]
Currently trying to add a new entry against an existing camera gives the message that the camera name must be unique. I'm not sure of the inner workings, but the request is to get rid of that camera name being unique constraint. If there is a need to keep something unique for settings storage, there could be an alternate field like a profile name, instead of attaching the uniqueness constraint on the camera name.

If we remove the camera name out of the constraint, that would allow for creating multiple entries for the same camera. Hope this gives some clarity.

**Comment by VorlonCD** [comment]
Ok, I believe this will work.   You should now be able to create as many cameras in AITOOL as you like and the names do not have to match what is in BI as long as you set the new BI Camera Name setting correctly.    It also lets you specific a custom mask file for each camera.   The variable [camera] is now expanded to "BI Camera Name" setting NOT the name of the camera in your AITOOL list.

![image](https://user-images.githubusercontent.com/7512113/103184590-d1f14680-4886-11eb-9350-692f2edc5799.png)

[AITOOL-VORLONCD.zip](https://github.com/VorlonCD/bi-aidetection/files/5745627/AITOOL-VORLONCD.zip)

Let me know how it works.

**Comment by balqosz** [comment]
Hi!

 Can the same image (same filename) be used to be scanned twice or more? i.e., I have one BI Camera Cam1 which outputs snapshot files named Cam1*. I would like to be able in AI Tool to create Cam1 and Cam1_b to perform different masks and alerts. Is that possible?
Thx

**Comment by tripp396** [comment]
> Hi!
> 
> Can the same image (same filename) be used to be scanned twice or more? i.e., I have one BI Camera Cam1 which outputs snapshot files named Cam1*. I would like to be able in AI Tool to create Cam1 and Cam1_b to perform different masks and alerts. Is that possible? Actually, I think that if the cam1 image is used by AI Tool on Cam1 camera setup, then it is discarded for a new detection in another AI Tool camera Cam1_b.
> 
> In other words, may I use twice the same image snapshot generated by only one camera in BI with two different AI Tool cameras?
> 
> Thx

I am also wondering this. I tried it and it doesn’t seem to work that way.

**Comment by VorlonCD** [comment]
@balqosz - Each image can only match one camera.   It would be tricky to change that since the history database is keyed to the full filename and each entry must be unique.

You can preform multiple actions on a single image detection's - Both TRIGGER URL and MQTT allow use of a | or ,  (Pipe or comma) between more than one item.  So you can have URL1 | URL2 or MQTTTOPIC1 | MQTTTOPIC2 to do multiple things.    Also, the **RUN EXTERNAL PROGRAM** can call a batch file or powershell script that preforms multiple actions.

With the version I posted above, you can have more than one camera that use the same BI camera.   So if you have BI set to create filenames that start with either CAM1-FIRSTPRESET or CAM1-SECONDPRESET (by using the &PRESET variable as mentioned by @ExcelsiorEyer ), you would create two cameras in AITOOL with different names, set each to CAM1 and the "**Input file begins with**" (prefix) set to CAM1-FIRSTPRESET and the second AITOOL camera set to CAM1-SECONDPRESET. 

If that doesnt help, give me a very specific example of what you want to accomplish.

**Comment by balqosz** [comment]
> @balqosz - Each image can only match one camera. It would be tricky to change that since the history database is keyed to the full filename and each entry must be unique.

![image](https://user-images.githubusercontent.com/24378221/103240143-43d09b00-492e-11eb-8eec-7b18f892e6da.png)

You can see in the snapshot one blue box and one red box. The blue box: I only keep alert history of people just walking by.
But in red box, I would prefer to flag the image and sound a short alarm.

I'm having success with it but using Ghost to copy the snapshot file and processing it in a second camera definition (virtual camera) in AI Tool.
But I think that the process of copying would not be needed for that purpose, if I could just scan the same image twice with different masks and alerts. It's not a priority, it's just an idea. I think one image from any camera should have the chance to be scanned twice or more, if needed.

**Comment by tripp396** [comment]
Similar use case for me. Can’t grab a screenshot since I’m not home but basically need to capture bottom third of frame, just people. But need to capture top 2/3 people and cars. Need the first situation to alert via telegram, and the other situation to not alert. 

Currently using an additional camera in BI to accomplish it

**Comment by balqosz** [comment]
Maybe with "external program" there is an alternative too, I must think about it... Thank you VorlonCD

**Comment by ExcelsiorEyer** [comment]
> Ok, I believe this will work. You should now be able to create as many cameras in AITOOL as you like and the names do not have to match what is in BI as long as you set the new BI Camera Name setting correctly. It also lets you specific a custom mask file for each camera. The variable [camera] is now expanded to "BI Camera Name" setting NOT the name of the camera in your AITOOL list.
> 
> ![image](https://user-images.githubusercontent.com/7512113/103184590-d1f14680-4886-11eb-9350-692f2edc5799.png)
> 
> [AITOOL-VORLONCD.zip](https://github.com/VorlonCD/bi-aidetection/files/5745627/AITOOL-VORLONCD.zip)
> 
> Let me know how it works.


Thanks @VorlonCD. Right now running into some issues.
I'm getting this error when trying to edit an existing camera to include the preset ID.
![image](https://user-images.githubusercontent.com/75917861/103243666-2f85a180-4920-11eb-8d77-4a3e03221a66.png)

Error message:
```
System.NullReferenceException: Object reference not set to an instance of an object.
   at AITool.Shell.CameraSave(Boolean SaveTo) in C:\Documents\Projects\_GIT\VorlonCD\bi-aidetection\src\UI\Shell.cs:line 2510
   at AITool.Shell.btnCameraSave_Click_1(Object sender, EventArgs e) in C:\Documents\Projects\_GIT\VorlonCD\bi-aidetection\src\UI\Shell.cs:line 2479
   at System.Windows.Forms.Control.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnMouseUp(MouseEventArgs mevent)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ButtonBase.WndProc(Message& m)
   at System.Windows.Forms.Button.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
```

**Comment by VorlonCD** [comment]
@ExcelsiorEyer - for now just delete/recreate the camera.

Also, it is assumed the 'prefix' or 'file starts with' ends at the first PERIOD of the filename.    So you cannot use REARCAMSD.12.    You have to tell BI to create a file that starts with REARCAMSD_12. or REARCAMSD-12.

**Comment by ExcelsiorEyer** [comment]
> @ExcelsiorEyer - for now just delete/recreate the camera.
> 
> Also, it is assumed the 'prefix' or 'file starts with' ends at the first PERIOD of the filename. So you cannot use REARCAMSD.12. You have to tell BI to create a file that starts with REARCAMSD_12. or REARCAMSD-12.

Ah okay. I'll recreate the entry on the tool. However, on the naming, looks like that's going to be tricky. BI requires the naming convention to have "CAM." or ".CAM". I can't get away with putting in a - or an _ in the naming. The last path component being the file name. 

Any thoughts?

![image](https://user-images.githubusercontent.com/75917861/103257287-3a5b2900-4956-11eb-9407-96f1f48f8c8f.png)

**Comment by VorlonCD** [comment]
This one should work a bit better.   For the "file starts with" setting, you should be able to use anything you like including  . or -, etc.    You dont have to, but If you use wildcard you can match any part of the filename.  Like *.PREFIX*

[AITOOL-VORLONCD.zip](https://github.com/VorlonCD/bi-aidetection/files/5758301/AITOOL-VORLONCD.zip)

**Comment by balqosz** [comment]
Hi all, Happy new year 2021!

 VorlonCD: Is it possible to change this message to Idle?

![image](https://user-images.githubusercontent.com/24378221/103430458-19d3de80-4ba3-11eb-8754-113d5949bec4.png)

**Comment by VorlonCD** [comment]
@balqosz - Seems like a big change.  I don't think I should do that without @gentlepumpkin approval.  :)

**Comment by balucanb** [comment]
@VorlonCD  Is this working now? I am using version="2.0.389.7670 I have a PTZ camera named ReolinkDWsd in BI. In AI-Tool I clicked on this camera, changed the "AI Tool Camera name" and "Input file begins with" name to ReolinkDW-001 (also tried ReolinkDW-001.) and clicked save. Is a new camera supposed to then show up in the camera box because it did not when I tried, it just renamed the camera, If I try to change the name back it tells me the camera already exists and if I change the 001 to the next preset 002 again it just changes the camera name in the camera box. Am I missing something? I have been reading the instuction above for a hour because I don't want to look like Spiny Lumpsucker, but I do not see what I am doing wrong.

**Comment by VorlonCD** [comment]
@balucanb - The ADD dialog box will only list actual configured blueiris cameras.   If you want to add a custom name you have to type it in the 'Or type new camera names...'.   So:
1) Rename AI TOOL camera from 'ReolinkDWsd' to 'ReolinkDW-001'.   Then change 'Input file begins with to 'ReolinkDW-001'
2) ADD a new camera by typing it in the ADD box - 'ReolinkDW-002'.   Then change the 'BI Camera name' in the new camera back to 'ReolinkDWsd'

**Comment by Yonny24** [comment]
If you have the same camera in the AI tool configured with unique names can you then process the same images in parallel, one with a custom model and the other with vision detection standard?

**Comment by VorlonCD** [comment]
No.

**Comment by balucanb** [comment]
> @balucanb - The ADD dialog box will only list actual configured blueiris cameras. If you want to add a custom name you have to type it in the 'Or type new camera names...'. So:
> 
> 1. Rename AI TOOL camera from 'ReolinkDWsd' to 'ReolinkDW-001'.   Then change 'Input file begins with to 'ReolinkDW-001'
> 2. ADD a new camera by typing it in the ADD box - 'ReolinkDW-002'.   Then change the 'BI Camera name' in the new camera back to 'ReolinkDWsd'

SO I got the camera(s) to rename correctly. Thanks. Do you just drop a marked up bmp into the lastimages folder in order to mask out areas based on the different presets (cameras)

**Comment by VorlonCD** [comment]
@balucanb - Just change the Camera > 'Custom Mask Filename' to something different per camera, and hit CUSTOM button to edit each one.
